textacy/corpora/reddit_reader.py

```diff
@@ -0,0 +1,132 @@
+"""
+Reddit Corpus Reader
+--------------------
+
+Stream a corpus of up to ~1.6 billion Reddit comments posted from October 2007
+until May 2015, as either plaintext strings or content + metadata dicts.
+
+.. code-block:: pycon
+
+    >>> rr = RedditReader('/path/to/RC_2015-01.bz2')
+    >>> for text in rr.texts(limit=5):  # plaintext comments
+    ...     print(text)
+    >>> for comment in rr.comments(min_len=100, limit=1):  # parsed comments
+    ...     print(comment.keys())
+    ...     print(comment['body'])
+    >>> rr = RedditReader(['/path/to/RC_2015-01.bz2', '/path/to/RC_2015-02.bz2'])
+
+Raw data is downloadable from https://archive.org/details/2015_reddit_comments_corpus.
+"""
+import datetime
+import json
+import re
+
+from textacy.compat import bzip_open, str
+from textacy.preprocess import normalize_whitespace
+
+
+REDDIT_LINK_RE = re.compile(r'\[([^]]+)\]\(https?://[^\)]+\)')
+
+
+class RedditReader(object):
+    """
+    Stream Reddit comments from standardized, compressed files on disk, either as
+    plaintext strings or dict documents with both text content and metadata.
+    Download the data from https://archive.org/details/2015_reddit_comments_corpus;
+    the files are named as `RC_YYYY-MM.bz2`.
+
+    Args:
+        paths (sequence[str] or str): name or names of reddit comment file(s) from
+            which to stream comments
+    """
+
+    def __init__(self, paths):
+        if isinstance(paths, str):
+            self.paths = (paths,)
+        else:
+            self.paths = tuple(paths)
+
+    def __iter__(self):
+        for path in self.paths:
+            try:
+                with bzip_open(path, mode='rt') as f:
+                    for line in f:
+                        yield json.loads(line)
+            except ValueError:  # Python 2 sucks and can't open bzip in text mode
+                with bzip_open(path, mode='r') as f:
+                    for line in f:
+                        yield json.loads(line)
+
+    def _clean_content(self, content):
+        # strip out link markup, e.g. [foo](http://foo.com)
+        content = REDDIT_LINK_RE.sub(r'\1', content)
+        # clean up basic HTML cruft
+        content = content.replace('&gt;', '>').replace('&lt;', '<')
+        # strip out text markup, e.g. * for bold text
+        content = content.replace('`', '').replace('*', '').replace('~', '')
+        # normalize whitespace
+        return normalize_whitespace(content)
+
+    def _parse_comment(self, comment):
+        # convert str/int timestamp fields into datetime objects
+        for key in ('created_utc', 'retrieved_on'):
+            try:
+                comment[key] = datetime.datetime.utcfromtimestamp(int(comment[key]))
```

Comment:
Trying to keep my package dependencies to a minimum... If the stdlib has exactly what I need, I might as well take advantage!